sklearn/impute.py

```diff
@@ -629,3 +632,262 @@ def fit_transform(self, X, y=None):
+                        self._sparse_fit(X, self.missing_values)
+        else:
+            self.uniques_, self.probas_ = self._dense_fit(X,
+                                                          self.missing_values)
+
+        return self
+
+    def _sparse_fit(self, X, missing_values):
+        """Fit the transformer on sparse data."""
+        mask_data = _get_mask(X.data, missing_values)
+        n_implicit_zeros = X.shape[0] - np.diff(X.indptr)
+
+        uniques = np.empty(X.shape[1], dtype=object)
+        probas = np.empty(X.shape[1], dtype=object)
+
+        for i in range(X.shape[1]):
+            column = X.data[X.indptr[i]:X.indptr[i+1]]
+            mask_column = mask_data[X.indptr[i]:X.indptr[i+1]]
+            column = column[~mask_column]
+
+            values, counts = self._get_values_counts(column)
+
+            # count implicit zeros
+            if n_implicit_zeros[i] > 0:
+                if 0 in values:
+                    counts[values == 0] += n_implicit_zeros[i]
+                else:
+                    values = np.append(values, 0)
+                    counts = np.append(counts, n_implicit_zeros[i])
+
+            if values.size > 0:
+                uniques[i] = values
+                if values.size == X.shape[0]:
+                    probas[i] = None
+                else:
+                    probas[i] = counts / counts.sum()
+            else:
+                uniques[i] = None
+                probas[i] = None
+
+        return uniques, probas
+
+    def _dense_fit(self, X, missing_values):
+        """Fit the transformer on dense data."""
+        mask = _get_mask(X, missing_values)
+
+        uniques = np.empty(X.shape[1], dtype=object)
+        probas = np.empty(X.shape[1], dtype=object)
+
+        for i in range(X.shape[1]):
+            column = X[~mask[:, i], i]
+            if column.size > 0:
+                uniques[i], counts = self._get_values_counts(column)
+                if uniques[i].size == column.size:
+                    probas[i] = None
+                else:
+                    probas[i] = counts / counts.sum()
+            else:
+                uniques[i] = None
+                probas[i] = None
+
+        return uniques, probas
+
+    def transform(self, X):
+        """Impute all missing values in X.
+
+        Parameters
+        ----------
+        X : {array-like, sparse matrix}, shape = [n_samples, n_features]
+            The input data to complete.
+        """
+        check_is_fitted(self, ("uniques_", "probas_"))
+
+        X = self._validate_input(X)
+
+        uniques = self.uniques_
+        probas = self.probas_
+
+        if X.shape[1] != uniques.shape[0]:
+            raise ValueError("X has %d features per sample, expected %d"
+                             % (X.shape[1], self.uniques_.shape[0]))
+
+        # Delete the invalid columns
+        valid_mask = self._vect_is_not_none(uniques)
+        valid_uniques = uniques[valid_mask]
```

Comment:
I probably shouldn't be bothered by it, but `valid_mask.all()` being True is by far expected to be the usual case... We could avoid the copying and indexing here by saying `valid_mask = Ellipsis` in the usual case...